salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,535 @@
+        # what format would be used in source manifests (most likely, there
+        # wouldn't be any date). We thus pick the Swagger `date-time` string
+        # format (see swagger.io/docs/specification/data-models/data-types/).
+        try:
+            return datetime.datetime.strptime(value, '%Y-%m-%dT%H:%M:%SZ')
+        except (TypeError, ValueError):
+            raise _type_error(value, expected='a date-time string')
+
+    dict_match = DICT_PATTERN.match(type_string)
+    if dict_match is not None:
+        if not isinstance(value, dict):
+            raise _type_error(value, expected='a dictionary')
+
+        value_type_str = dict_match.group('value_type')
+        return {
+            key: _cast_value(val, value_type_str)
+            for key, val in value.items()
+        }
+
+    list_match = LIST_PATTERN.match(type_string)
+    if list_match is not None:
+        if not isinstance(value, list):
+            raise _type_error(value, expected='a list')
+
+        value_type_str = list_match.group('value_type')
+        return [_cast_value(val, value_type_str) for val in value]
+
+    try:
+        model = getattr(k8s_client.models, type_string)
+    except AttributeError:
+        # This should never happen, otherwise this function should get updated
+        raise ValueError(
+            'Unknown type string provided: {}.'.format(type_string)
+        )
+
+    if not isinstance(value, dict):
+        raise _type_error(
+            value,
+            expected='a dict to cast as a "{}"'.format(model.__name__)
+        )
+
+    return _build_standard_object(model, value)
+
+
+def _type_error(value, expected):
+    return TypeError(
+        'Expected {}, received: {} (type: {})'.format(
+            expected, value, type(value)
+        )
+    )
+
+
+def validate_manifest(manifest):
+    """Ensure a Kubernetes object is strictly conformant to its OpenAPI schema.
+
+    This method relies on the K8s client models, which are generated from said
+    schema. The approach is to cast the original `manifest` dict into such
+    models, and then obtain the dict representation from the model back again,
+    ensuring that nothing is lost from the original manifest (the conversion
+    also checks that no "unsupported" field is provided).
+
+    For CustomResources, this only checks that the apiVersion and kind are
+    registered in this file.
+    """
+    obj = convert_manifest_into_object(manifest)
+    result = obj.to_dict()
+
+    # We change camelCase into snake_case for all keys in the source
+    # manifest, as is done by the `kubernetes` client.
+    desired = _cast_dict_keys(manifest, key_cast=_camel_to_snake)
+
+    # We only check fields provided in the source, since the conversion
+    # may add empty or default values for optional fields
+    dictdiff = recursive_diff(desired, result)
+    return not dictdiff.removed() and not dictdiff.changed()
+
+
+def _cast_dict_keys(data, key_cast):
```

Comment:
So `_cast_dict_keys` handles `list`? That's confusing.

Reply:
Even handles neither `list` nor `dict` as a passthrough. It warrants a docstring, agreed that it can be confusing.